_posts/2018-09-14-getkw-01.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: post
+title: Redesigning Getkw _or_ reinventing the wheel: 0. State of affairs and requirements
+author: Roberto Di Remigio
+summary: Redesigning Getkw
+status: wip
+---
+
+This is part 0 of a series of posts on designing a general input parsing library.
+By _general input parsing library_ we mean that the format of
+the input is not fixed by the library itself. 
+
+In this post we will cover the state of affairs and lay out in detail the
+requirements for our new library.
+
+## State of affairs
+
+The [Getkw library] was designed with the goal of being _general_.
+The principle of operation of Getkw is:
+1. Give me an input file.
+2. I'll parse it and check whether it's a legal input.
+4. If it is _not legal_, bye, bye.
+3. If it is, I'll dump a machine-readable file to disk. The file has all
+   possible input parameters whether at their default value or the value passed
+   by the user.
+4. A reader for your language of choice (currently C, C++, and Fortran) will
+   gobble the machine-readable file _et voilà_ you have your input parameters.
+
+The structure of the input revolves around the concept of keywords and sections.
+**Keywords** are the basic entity, **sections** collect keywords, or other
+sections. Loosely speaking, keywords are the bottom layer, while sections can be
+recursive. We'll detail this point later.
+
+### How are sanity checks performed?
+
+Each section gets a callback function to perform checks on its contents when the
+section is parsed.
+
+### Strengths
+
+a. Invalid inputs are blocked before CPU cycles are consumed.
+b. No regexes are used, rather the input is treated as a formal grammar with its
+   [Backus-Naur form].
+
+### Weaknesses
+
+a. Getkw is:
+   - A Python module. This uses `pyparsing.py` to define the grammar for all
```

Comment:
But why is it a weakness that getkw is a python module and uses `pyparsing.py`? Not questioning this statement but the justification may not be clear to the reader - it is not clear to me.

Reply:
It's the combination of Python/compiled languages that is a weakness, IMO. Though I agree that it is not very clear now, perhaps